homeassistant/components/sonos/helpers.py

```diff
@@ -1,43 +1,69 @@
 """Helper methods for common tasks."""
 from __future__ import annotations
 
-from collections.abc import Callable
-import functools as ft
 import logging
-from typing import Any
+from typing import TYPE_CHECKING, Any, Callable, TypeVar, cast
 
 from soco.exceptions import SoCoException, SoCoUPnPException
 
 from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.dispatcher import dispatcher_send
+
+from .const import SONOS_SPEAKER_ACTIVITY
+from .exception import SpeakerUnavailable
+
+if TYPE_CHECKING:
+    from .entity import SonosEntity
+    from .speaker import SonosSpeaker
 
 UID_PREFIX = "RINCON_"
 UID_POSTFIX = "01400"
 
+WrapFuncType = TypeVar("WrapFuncType", bound=Callable[..., Any])
+
 _LOGGER = logging.getLogger(__name__)
 
 
-def soco_error(errorcodes: list[str] | None = None) -> Callable:
+def soco_error(
+    errorcodes: list[str] | None = None, raise_on_err: bool = True
+) -> Callable:
     """Filter out specified UPnP errors and raise exceptions for service calls."""
 
-    def decorator(funct: Callable) -> Callable:
+    def decorator(funct: WrapFuncType) -> WrapFuncType:
         """Decorate functions."""
 
-        @ft.wraps(funct)
-        def wrapper(*args: Any, **kwargs: Any) -> Any:
+        def wrapper(self: SonosSpeaker | SonosEntity, *args: Any, **kwargs: Any) -> Any:
             """Wrap for all soco UPnP exception."""
             try:
-                return funct(*args, **kwargs)
+                result = funct(self, *args, **kwargs)
+            except SpeakerUnavailable:
+                return None
             except (OSError, SoCoException, SoCoUPnPException) as err:
                 error_code = getattr(err, "error_code", None)
                 function = funct.__name__
                 if errorcodes and error_code in errorcodes:
                     _LOGGER.debug(
                         "Error code %s ignored in call to %s", error_code, function
                     )
-                    return
-                raise HomeAssistantError(f"Error calling {function}: {err}") from err
+                    return None
+
+                # Prefer the entity_id if available, zone name as a fallback
+                # Needed as SonosSpeaker instances are not entities
+                zone_name = getattr(self, "speaker", self).zone_name
+                target = getattr(self, "entity_id", zone_name)
```

Comment:
Thanks.  I couldn't think of a way to make this cleaner without a whole lot more code.  Maybe we will come up with something in the future.